components/higher-order/with-dragging/index.js

```diff
@@ -0,0 +1,155 @@
+/**
+ * WordPress Dependencies
+ */
+import { Component } from '@wordpress/element';
+
+/**
+ * Internal Dependencies
+ */
+import './style.scss';
+
+const dragImageClass = 'components-with-dragging__invisible-drag-image';
+const cloneWrapperClass = 'components-with-dragging__clone';
+const cloneHeightTransformationBreakpoint = 700;
+const clonePadding = 20;
+const context = {
+	cloneNodeId: null,
+	elementId: null,
+	cursorTop: null,
+	cursorLeft: null,
+};
+
+const withDragging = ( OriginalComponent ) => {
+	class Draggable extends Component {
+		constructor() {
+			super( ...arguments );
+			this.onDragStart = this.onDragStart.bind( this );
+			this.onDragOver = this.onDragOver.bind( this );
+			this.onDragEnd = this.onDragEnd.bind( this );
+		}
+
+		/**
+		 * Removes the element clone, resets cursor, and removes drag listener.
+		 * @param  {Object} event     The non-custom DragEvent.
+		 */
+		onDragEnd( event ) {
+			const element = document.getElementById( context.elementId );
+			const cloneWrapper = document.getElementById( context.cloneNodeId );
+
+			if ( element && cloneWrapper ) {
+				// Remove clone.
+				element.parentNode.removeChild( cloneWrapper );
+				context.elementId = null;
+				context.cloneNodeId = null;
+			}
+
+			// Reset cursor.
+			document.body.classList.remove( 'dragging' );
+			document.removeEventListener( 'dragover', this.onDragOver );
+			event.stopPropagation();
+			event.preventDefault();
+		}
+
+		/*
+		 * Updates positioning of element clone based on mouse movement during dragging.
+		 * @param  {Object} event     The non-custom DragEvent.
+		 */
+		onDragOver( event ) {
+			const cloneWrapper = document.getElementById( context.cloneNodeId );
+
+			cloneWrapper.style.top =
+				`${ parseInt( cloneWrapper.style.top, 10 ) + event.clientY - context.cursorTop }px`;
+			cloneWrapper.style.left =
+				`${ parseInt( cloneWrapper.style.left, 10 ) + event.clientX - context.cursorLeft }px`;
+
+			// Update cursor coordinates.
+			context.cursorLeft = event.clientX;
+			context.cursorTop = event.clientY;
+		}
+
+		/**
+		 *  - Clones the current element and spawns clone over original element.
+		 *  - Adds a fake temporary drag image to avoid browser defaults.
+		 *  - Sets transfer data.
+		 *  - Adds dragover listener.
+		 * @param  {Object} event					The non-custom DragEvent.
+		 * @param  {string} elementId				The HTML id of the element to be dragged.
+		 * @param  {Object} transferData			The data to be set to event's dataTransfer and will be made accessible in any later drop logic.
+		 * @param  {string} transferData.rootUID	A unique identifier of the element's root.
```

Comment:
Reminder to remove these inner property docs. It is largely irrelevant to onDragStart what `transferData` is passed here. All it does is set to the event to be handled by other handlers.